ON-14945: ON-15027: Separate build and deploy steps to support using pre-built images in air-gapped clusters. (#24)
Xilinx-CNS/kubernetes-onload: README.md

```
Remove any outstanding manually with `oc delete image`. (Not providing any automated invocation to prevent removal of the false-positive images.)

## Deploying artifacts into an airgapped cluster

### Images required

> This is a list of the images that are currently needed as reported by podman.

Reviewer: ```suggestion
This is a list of the images that are currently needed as reported by podman:
```

I know it's really nit-picking but I can't ignore typography inconsistency! We tend to use : (colon mark) for referring to the code snippets. If you disagree, then choose any approach and stick to it here and elsewhere like "Podman login". 👍

> ```

Reviewer: ```suggestion
```console
```
nit: I recently learnt that \`\`\`console is a better way for a multiline snippets like that (with one command and the rest is output)!

Reviewer 2: This can be part of ON-14978.

which will write the image into `images/sfc.tar`. Then use `podman load` to
load the image from the written file:
```
$ sudo podman load -i images.tar
```

### Podman login

In order to be able to pull/push images into the cluster's image registry you
must log in with podman.

1. log in to openshift cluster
```
$ oc login -u kubeadmin -p PASSWORD
```
2. Get the name of the image registry
```
oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}'
```
This can be stored in a local variable
```
REGHOST=`oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}'`
```
3. Actually login to the image registry with podman
```
sudo podman login -u kubeadmin -p $(oc whoami -t) ${REGHOST} --tls-verify=false
```

#### A Note on `--tls-verify=false`

Without doing any extra steps podman will complain about the certificates used
by the image registry. For development purposes `--tls-verify=false` can be
used to bypass this issue. Production environments shall follow internal
certificate handling procedures to secure podman access.

### Pushing the images to the cluster's image-registry

For each image you want to push run the command:
```

> $sudo podman push REGISTRY/OPENSHIFT_NAMESPACE/IMAGE_NAME:TAG --tls-verify=false

Reviewer: ```suggestion
$ sudo podman push REGISTRY/OPENSHIFT_NAMESPACE/IMAGE_NAME:TAG --tls-verify=false
```

```
Remove any outstanding manually with `oc delete image`. (Not providing any automated invocation to prevent removal of the false-positive images.)

## Deploying artifacts into an airgapped cluster

> ### Images needed

Reviewer 2: ```suggestion
### Images required
```

This is a list of the images that are currently needed as reported by podman.
```
$ sudo podman images --format "table {{.Repository}} {{.Tag}}"
REPOSITORY                                                                                           TAG
default-route-openshift-image-registry.apps.test.kube.test/default/onload-sfnettest                  sfnettest-1.6.0-rc1
default-route-openshift-image-registry.apps.test.kube.test/sfptpd/sfptpd                             git-ab881b3
default-route-openshift-image-registry.apps.test.kube.test/onload-clusterlocal/onload-module         git27b3826-4.18.0-372.49.1.el8_6.x86_64
default-route-openshift-image-registry.apps.test.kube.test/onload-clusterlocal/onload-device-plugin  latest
default-route-openshift-image-registry.apps.test.kube.test/onload-clusterlocal/onload-user           git27b3826
default-route-openshift-image-registry.apps.test.kube.test/openshift-kmm/sfc-module                  git27b3826-4.18.0-372.49.1.el8_6.x86_64
```

### Getting the images

> The images can either be built using podman directly (untested) or using a
> non-airgapped cluster.

Reviewer 2: Restricted networks which have local Git and Container Registry hosting can still use HTTP-based build methods.

```suggestion
The images can either be built using podman directly (currently unsupported) or on a
cluster with access to a git web host, either github.com or a locally hosted clone of <named repos>.
```

> #### Pulling images from a cluster

Reviewer 2: OpenShift sysadmins will be highly proficient in image handling so will not require an introduction to podman. This section would be better served by a ready-to-go script covering all the necessary images.

Author: I think I'm going to leave this as is for the time being since this document is useful to members of our team who aren't proficient with image handling.

Reviewer: How about moving the `podman` sections into a separate README file? The top-level one becomes overwhelming. I bet we should keep it short and friendly!

If you think the main consumers are going to be our team members (who are highly skilled! 😜), how about moving these bits into the internal knowledge base like Confluence?

```

`podman save` can be used to save the into a local file, e.g:
```
$ sudo podman save -o images/sfc.tar default-route-openshift-image-registry.apps.test.kube.test/openshift-kmm/sfc-module:git27b3826-4.18.0-372.49.1.el8_6.x86_64
```
which will write the image into `images/sfc.tar`. Then use `podman load` to
load the image from the written file:
```
$ sudo podman load -i images.tar
```

### Podman login

In order to be able to pull/push images into the cluster's image registry you
must log in with podman.

1. log in to openshift cluster
```
$ oc login -u kubeadmin -p PASSWORD
```
2. Get the name of the image registry
```
oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}'
```
This can be stored in a local variable
```
REGHOST=`oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}'`
```
3. Actually login to the image registry with podman
```
sudo podman login -u kubeadmin -p $(oc whoami -t) ${REGHOST} --tls-verify=false
```

#### A Note on `--tls-verify=false`

Without doing any extra steps podman will complain about the certificates used
by the image registry. For development purposes `--tls-verify=false` can be

> used to bypass this issue. A proper solution would be to extract the
> certificates and pass these to podman.

Reviewer 2: ```suggestion
used to bypass this issue. Production environments shall follow internal certificate handling procedures to secure podman access.
```

### Set up the imagestreams and namespaces

```
$ oc create -f sfc/imagestream/imagestream.yaml
$ oc create -f onload/imagestream/

$ oc create -f sfptpd/sfptpd-imagestream.yaml

$ oc create -f examples/sfnettest/0000-imagestream.yaml
```

### Pushing the images to the cluster's image-registry

For each image you want to push run the command:
```
$sudo podman push REGISTRY/OPENSHIFT_NAMESPACE/IMAGE_NAME:TAG --tls-verify=false
```

### Deploying sfc driver

#### KMM deployment
First remove the in-tree driver from the worker nodes
```
# rmmod sfc
```

> This should allow the kmm to install the kernel module without issue.

Reviewer 2: If this command is required it should be added to `sfc/mco/99-sfc-machineconfig.bu` rather than left as an exercise for the user.

Author: Removing `sfc` and `sfc_driverlink` is already handled in [`sfc-replace-kernel-module.sh#L9`](https://github.com/Xilinx-CNS/kubernetes-onload/blob/e556d4d5186cca334ccf56424263f658bd6e6036/sfc/mco/sfc-replace-kernel-module.sh#L9).

This is only necessary since we are using KMM v1.0 (which is the only one I can find on the operator hub). If we were using KMM v1.1 there is the `inTreeModuleToRemove` field which would handle the removal for us.

Reviewer 2: Good point, this is useful information to have in that section of the documentation. 

~~~suggestion
#### KMM Module deployment

If KMM v1.0 is used, the in-tree driver must be removed from the worker nodes
before deployment. This is a one-time action which is automated by the
MachineConfig component upon reboot. In KMM v1.1 onwards, add 
`inTreeModuleToRemove: sfc` to the `sfc` Module CR.

```sh
rmmod sfc
```
~~~

Author: Your suggestions looks good, I've added it as a separate commit
